### PR TITLE
NEXT-00000 - Prevent product being loaded in own cross selling product streams

### PR DESCRIPTION
**Pull request by lacknere**
### 1. Why is this change necessary?
Products should not be loaded/displayed in their own cross sellings.

### 2. What does this change do, exactly?
* Added filter to product stream criteria in `ProductCrossSellingRoute::loadByStream`.

### 3. Describe each step to reproduce the issue or behaviour.
Create a product stream containing the product you are going to add a cross selling to > add a cross selling using the created product stream to the product

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: changelog/_unreleased/2024-09-18-prevent-product-being-loaded-in-own-cross-selling-product-streams.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+title: Prevent product being loaded in own cross selling product streams
+issue: NEXT-00000
+author: Elias Lackner
+author_email: lackner.elias@gmail.com
+author_github: @lacknere
+---
+# Core
+* Added filter to product stream criteria in `ProductCrossSellingRoute::loadByStream`.
```
Comment (phuoccx) [inline changelog/_unreleased/2024-09-18-prevent-product-being-loaded-in-own-cross-selling-product-streams.md]: Please describe more details(purpose) of the filter that you added

File: tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php [hunk 2]
```diff
@@ -500,11 +500,40 @@ static function (ProductCrossSellingIdsCriteriaEvent $event) use ($productReposi
...
+
+        $this->productRepository->create([$productData], $this->salesChannelContext->getContext());
+
+        $result = $this->route->load($productId, new Request(), $this->salesChannelContext, new Criteria())
+            ->getResult();
+
+        $element = $result->first();
+
+        static::assertNotNull($element);
+        static::assertNotContains($productId, $element->getProducts()->getIds());
```
Comment (phuoccx) [inline tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]: Could you please add more detailed assertions here?
example:
We have a product stream with more than one product inside
The result should be to return a collection of product streams without productId!

Comment (lacknere) [reply tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]: I've added an assertion for the product count.
The assertion with the productId is already there, no?
Also, I've removed the cross selling limit which made the testcase irrelevant.

Comment (phuoccx) [reply tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]:
```php 
static::assertEquals(5, $element->getProducts()->count()); 
```
That's what I want. 
Everything looks good to me. However, we have failed the UT. Please check it.

Comment (mitelg) [reply tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]: please use assertCount :see_no_evil:

Comment (lacknere) [reply tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]: Yeah, the failing testcase was expecting the product to be loaded into its own cross selling 😄 
I've removed the product from the expected cross selling result.

Comment (lacknere) [reply tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]: @mitelg Sure, I've changed it for the whole test class.

Comment (lacknere) [reply tests/integration/Core/Content/Product/SalesChannel/CrossSelling/CrossSellingRouteTest.php]: I've added some notNull assertions to fix the phpstan errors as well.

**Comment by leduc92** [comment]
Hi @lacknere, we have merged your pull request. 🎉  Thanks for your contribution.

**Pull request closed without merging**